### PR TITLE
to fix v1.5.8 Usage return 

### PR DESCRIPTION
**Pull request by nasa1024**
In the pr on [215](https://github.com/sashabaranov/go-openai/pull/215). I found that I used the wrong way to add Usage. when `stream = true` will not  return `Usage` filed.
usage
`Usage` should be calculated by the user himself.

### CODE REVIEW

**Comment by sashabaranov** [comment]
@nasa1024 hey, you meant to remove this line in this PR? (see diff)

**Comment by nasa1024** [comment]
here is the right way to caculator `usage`

```bash
go get github.com/pkoukk/tiktoken-go
```
here is example to get `usage`

```go
package main

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"

	"github.com/pkoukk/tiktoken-go"
	openai "github.com/sashabaranov/go-openai"
)

func main() {
	chat()
	stream()
}

func chat() {
	client := openai.NewClient("")
	resp, err := client.CreateChatCompletion(
		context.Background(),
		openai.ChatCompletionRequest{
			Model:     openai.GPT3Dot5Turbo,
			MaxTokens: 20,
			Messages: []openai.ChatCompletionMessage{
				{
					Role:    openai.ChatMessageRoleAssistant,
					Content: "你好，我叫小梦",
				},
				{
					Role:    openai.ChatMessageRoleUser,
					Content: "什么是算法稳定币？",
				},
			},
		},
	)

	if err != nil {
		fmt.Printf("ChatCompletion error: %v\n", err)
		return
	}

	fmt.Printf("response: %#v\n", resp)
}

func stream() {
	c := openai.NewClient("")
	ctx := context.Background()
	var response openai.ChatCompletionResponse
	var completionText string

	req := openai.ChatCompletionRequest{
		Model:     openai.GPT3Dot5Turbo,
		MaxTokens: 20,
		Messages: []openai.ChatCompletionMessage{
			{
				Role:    openai.ChatMessageRoleAssistant,
				Content: "你好，我叫小梦",
			},
			{
				Role:    openai.ChatMessageRoleUser,
				Content: "什么是算法稳定币？",
			},
		},
		Stream: true,
	}
	stream, err := c.CreateChatCompletionStream(ctx, req)
	if err != nil {
		fmt.Printf("ChatCompletionStream error: %v\n", err)
		return
	}
	defer stream.Close()

	fmt.Printf("Stream response: ")
	for {
		response, err := stream.Recv()
		if errors.Is(err, io.EOF) {
			fmt.Println("\nStream finished")
			break
		}

		if err != nil {
			fmt.Printf("\nStream error: %v\n", err)
			return
		}

		completionText += response.Choices[0].Delta.Content
		fmt.Printf(response.Choices[0].Delta.Content)
	}

	str, err := json.Marshal(req)
	if err != nil {
		err = fmt.Errorf(": %v", err)
		return
	}
	fmt.Printf("Request: %s", string(str))

	response.Usage.PromptTokens = NumTokensFromMessages(req.Messages, openai.GPT3Dot5Turbo)
	response.Usage.CompletionTokens = getTokenByModel(completionText, openai.GPT3Dot5Turbo)
	response.Usage.TotalTokens = response.Usage.PromptTokens + response.Usage.CompletionTokens
	fmt.Printf("response: %#v\n", response)
}

// getTokenByEncoding
func getTokenByModel(text string, model string) (num_tokens int) {

	tkm, err := tiktoken.EncodingForModel(model)
	if err != nil {
		err = fmt.Errorf(": %v", err)
		return
	}

	token := tkm.Encode(text, nil, nil)

	return len(token)
}

// NumTokensFromMessages
func NumTokensFromMessages(messages []openai.ChatCompletionMessage, model string) (num_tokens int) {
	tkm, err := tiktoken.EncodingForModel(model)
	if err != nil {
		err = fmt.Errorf("EncodingForModel: %v", err)
		fmt.Println(err)
		return
	}

	var tokens_per_message int
	var tokens_per_name int
	if model == "gpt-3.5-turbo-0301" || model == "gpt-3.5-turbo" {
		tokens_per_message = 4
		tokens_per_name = -1
	} else if model == "gpt-4-0314" || model == "gpt-4" {
		tokens_per_message = 3
		tokens_per_name = 1
	} else {
		fmt.Println("Warning: model not found. Using cl100k_base encoding.")
		tokens_per_message = 3
		tokens_per_name = 1
	}

	for _, message := range messages {
		num_tokens += tokens_per_message
		num_tokens += len(tkm.Encode(message.Content, nil, nil))
		if message.Name != "" {
			num_tokens += tokens_per_name
		}
	}
	num_tokens += 3
	return num_tokens
}

```

I'm sorry for increase in your workload and the error of the commit history

**Comment by nasa1024** [comment]
@collinvandyck
there is have prompt token no equal 

```bash
response: openai.ChatCompletionResponse{ID:"chatcmpl-71EycXmMs6ofZi4IdbUwEaYP1T8iR", Object:"chat.completion", Created:1680530254, Model:"gpt-3.5-turbo-0301", Choices:[]openai.ChatCompletionChoice{openai.ChatCompletionChoice{Index:0, Message:openai.ChatCompletionMessage{Role:"assistant", Content:"算法稳定币是一种数字货币，与其他稳定", Name:""}, FinishReason:"length"}}, Usage:openai.Usage{PromptTokens:35, CompletionTokens:20, TotalTokens:55}}
Stream response: 算法稳定币是一种基于算法的数字货币，
Stream finished
Request: {"model":"gpt-3.5-turbo","messages":[{"role":"assistant","content":"你好，我叫小梦"},{"role":"user","content":"什么是算法稳定币？"}],"max_tokens":20,"stream":true}response: openai.ChatCompletionResponse{ID:"", Object:"", Created:0, Model:"", Choices:[]openai.ChatCompletionChoice(nil), Usage:openai.Usage{PromptTokens:33, CompletionTokens:20, TotalTokens:53}}
```
maybe we need some discuss more

**Comment by collinvandyck** [comment]
@nasa1024 I believe that @sashabaranov is saying that because usage does not come back in the actual HTTP response for stream completions, that adding `omitempty` is not really the right approach, since it will always be empty. I think it should be removed altogether.

Also thanks for the call out on how to calculate tokens manually. This seems like a cool thing to have in go-openai -- the Recv code could possibly use this and just calculate the usage for you based on the model that generated the stream response.

**Comment by sashabaranov** [comment]
@nasa1024 As @collinvandyck correctly mentioned — please just remove the `Usage` field from the response, it would not be there anyway.

File: chat_stream.go [hunk 1]
```diff
@@ -21,7 +21,7 @@ type ChatCompletionStreamResponse struct {
 	Created int64                        `json:"created"`
 	Model   string                       `json:"model"`
 	Choices []ChatCompletionStreamChoice `json:"choices"`
-	Usage   Usage                        `json:"usage"`
+	Usage   Usage                        `json:"usage,omitempty"`
```
Comment (sashabaranov) [inline chat_stream.go:24]:
```suggestion
```

Comment (nasa1024) [reply chat_stream.go:24]: ok

**Review by nasa1024 — COMMENTED** [review]
remove the Usage field from the response

**Review by sashabaranov — APPROVED** [review]

**Comment by kaijietti** [comment]
> ```go
> for _, message := range messages {
> 		num_tokens += tokens_per_message
> 		num_tokens += len(tkm.Encode(message.Content, nil, nil))
> 		if message.Name != "" {
> 			num_tokens += tokens_per_name
> 		}
> 	}
> ```

这里应该还漏掉了：
num_tokens += len(tkm.Encode(message.Role, nil, nil))
以及
num_tokens += len(tkm.Encode(message.Name, nil, nil))

**Comment by valiner** [comment]
> here is the right way to caculator `usage`
> 
> ```shell
> go get github.com/pkoukk/tiktoken-go
> ```
> 
> here is example to get `usage`
> 
> ```go
> package main
> 
> import (
> 	"context"
> 	"encoding/json"
> 	"errors"
> 	"fmt"
> 	"io"
> 
> 	"github.com/pkoukk/tiktoken-go"
> 	openai "github.com/sashabaranov/go-openai"
> )
> 
> func main() {
> 	chat()
> 	stream()
> }
> 
> func chat() {
> 	client := openai.NewClient("")
> 	resp, err := client.CreateChatCompletion(
> 		context.Background(),
> 		openai.ChatCompletionRequest{
> 			Model:     openai.GPT3Dot5Turbo,
> 			MaxTokens: 20,
> 			Messages: []openai.ChatCompletionMessage{
> 				{
> 					Role:    openai.ChatMessageRoleAssistant,
> 					Content: "你好，我叫小梦",
> 				},
> 				{
> 					Role:    openai.ChatMessageRoleUser,
> 					Content: "什么是算法稳定币？",
> 				},
> 			},
> 		},
> 	)
> 
> 	if err != nil {
> 		fmt.Printf("ChatCompletion error: %v\n", err)
> 		return
> 	}
> 
> 	fmt.Printf("response: %#v\n", resp)
> }
> 
> func stream() {
> 	c := openai.NewClient("")
> 	ctx := context.Background()
> 	var response openai.ChatCompletionResponse
> 	var completionText string
> 
> 	req := openai.ChatCompletionRequest{
> 		Model:     openai.GPT3Dot5Turbo,
> 		MaxTokens: 20,
> 		Messages: []openai.ChatCompletionMessage{
> 			{
> 				Role:    openai.ChatMessageRoleAssistant,
> 				Content: "你好，我叫小梦",
> 			},
> 			{
> 				Role:    openai.ChatMessageRoleUser,
> 				Content: "什么是算法稳定币？",
> 			},
> 		},
> 		Stream: true,
> 	}
> 	stream, err := c.CreateChatCompletionStream(ctx, req)
> 	if err != nil {
> 		fmt.Printf("ChatCompletionStream error: %v\n", err)
> 		return
> 	}
> 	defer stream.Close()
> 
> 	fmt.Printf("Stream response: ")
> 	for {
> 		response, err := stream.Recv()
> 		if errors.Is(err, io.EOF) {
> 			fmt.Println("\nStream finished")
> 			break
> 		}
> 
> 		if err != nil {
> 			fmt.Printf("\nStream error: %v\n", err)
> 			return
> 		}
> 
> 		completionText += response.Choices[0].Delta.Content
> 		fmt.Printf(response.Choices[0].Delta.Content)
> 	}
> 
> 	str, err := json.Marshal(req)
> 	if err != nil {
> 		err = fmt.Errorf(": %v", err)
> 		return
> 	}
> 	fmt.Printf("Request: %s", string(str))
> 
> 	response.Usage.PromptTokens = NumTokensFromMessages(req.Messages, openai.GPT3Dot5Turbo)
> 	response.Usage.CompletionTokens = getTokenByModel(completionText, openai.GPT3Dot5Turbo)
> 	response.Usage.TotalTokens = response.Usage.PromptTokens + response.Usage.CompletionTokens
> 	fmt.Printf("response: %#v\n", response)
> }
> 
> // getTokenByEncoding
> func getTokenByModel(text string, model string) (num_tokens int) {
> 
> 	tkm, err := tiktoken.EncodingForModel(model)
> 	if err != nil {
> 		err = fmt.Errorf(": %v", err)
> 		return
> 	}
> 
> 	token := tkm.Encode(text, nil, nil)
> 
> 	return len(token)
> }
> 
> // NumTokensFromMessages
> func NumTokensFromMessages(messages []openai.ChatCompletionMessage, model string) (num_tokens int) {
> 	tkm, err := tiktoken.EncodingForModel(model)
> 	if err != nil {
> 		err = fmt.Errorf("EncodingForModel: %v", err)
> 		fmt.Println(err)
> 		return
> 	}
> 
> 	var tokens_per_message int
> 	var tokens_per_name int
> 	if model == "gpt-3.5-turbo-0301" || model == "gpt-3.5-turbo" {
> 		tokens_per_message = 4
> 		tokens_per_name = -1
> 	} else if model == "gpt-4-0314" || model == "gpt-4" {
> 		tokens_per_message = 3
> 		tokens_per_name = 1
> 	} else {
> 		fmt.Println("Warning: model not found. Using cl100k_base encoding.")
> 		tokens_per_message = 3
> 		tokens_per_name = 1
> 	}
> 
> 	for _, message := range messages {
> 		num_tokens += tokens_per_message
> 		num_tokens += len(tkm.Encode(message.Content, nil, nil))
> 		if message.Name != "" {
> 			num_tokens += tokens_per_name
> 		}
> 	}
> 	num_tokens += 3
> 	return num_tokens
> }
> ```
> 
> I'm sorry for increase in your workload and the error of the commit history

thx

**Comment by nicewook** [comment]
At the moment(6/27), Counting token is slightly changed. 
```
// below link may not work on Chrome(error: Unable to render code block)
// then, use FireFox
// OpenAI Cookbook: https://github.com/openai/openai-cookbook/blob/main/examples/How_to_count_tokens_with_tiktoken.ipynb
func NumTokensFromMessages(messages []openai.ChatCompletionMessage, model string) (numTokens int) {
	tkm, err := tiktoken.EncodingForModel(model)
	if err != nil {
		err = fmt.Errorf("encoding for model: %v", err)
		log.Println(err)
		return
	}

	var tokensPerMessage, tokensPerName int

	if model == "gpt-3.5-turbo-0613" ||
		model == "gpt-3.5-turbo-16k-0613" ||
		model == "gpt-4-0314" ||
		model == "gpt-4-32k-0314" ||
		model == "gpt-4-0613" ||
		model == "gpt-4-32k-0613" {
		tokensPerMessage = 3
		tokensPerName = -1
	} else if model == "gpt-3.5-turbo-0301" {
		tokensPerMessage = 4 // every message follows <|start|>{role/name}\n{content}<|end|>\n
		tokensPerName = -1   // if there's a name, the role is omitted
	} else if model == "gpt-3.5-turbo" {
		log.Println("warning: gpt-3.5-turbo may update over time. Returning num tokens assuming gpt-3.5-turbo-0613.")
		return NumTokensFromMessages(messages, "gpt-3.5-turbo-0613")
	} else if model == "gpt-4" {
		log.Println("warning: gpt-4 may update over time. Returning num tokens assuming gpt-4-0613.")
		return NumTokensFromMessages(messages, "gpt-4-0613")
	} else {
		err := errors.New("warning: model not found. Using cl100k_base encoding")
		log.Println(err)
		return
	}

	for _, message := range messages {
		numTokens += tokensPerMessage
		numTokens += len(tkm.Encode(message.Content, nil, nil))
		numTokens += len(tkm.Encode(message.Role, nil, nil))
		numTokens += len(tkm.Encode(message.Name, nil, nil))
		if message.Name != "" {
			numTokens += tokensPerName
		}
	}
	numTokens += 3 // every reply is primed with <|start|>assistant<|message|>
	return numTokens
}
```

**Pull request merged**